File: docs-examples/example-java/src/test/java/io/micronaut/mqtt/docs/custom/annotation/CorrelationAnnotationBinder.java
package io.micronaut.mqtt.docs.custom.annotation;

// tag::imports[]
import io.micronaut.core.type.Argument;
import io.micronaut.mqtt.bind.AnnotatedMqttBinder;
import io.micronaut.core.convert.ArgumentConversionContext;
import io.micronaut.core.convert.ConversionService;
import io.micronaut.mqtt.v5.bind.MqttV5BindingContext;

import javax.inject.Singleton;
import java.util.Optional;
// end::imports[]

import io.micronaut.context.annotation.Requires;

@Requires(property = "spec.name", value = "CorrelationSpec")
// tag::clazz[]
@Singleton // <1>
public class CorrelationAnnotationBinder implements AnnotatedMqttBinder<MqttV5BindingContext, Correlation> { // <2>

    private final ConversionService<?> conversionService;

    public CorrelationAnnotationBinder(ConversionService<?> conversionService) { // <3>
        this.conversionService = conversionService;
    }

    @Override
    public Class<Correlation> getAnnotationType() {
        return Correlation.class;
    }

    @Override
    public void bindTo(MqttV5BindingContext context, Object value, Argument<Object> argument) {
        context.getProperties().setCorrelationData((byte[]) value); // <4>
    }

    @Override
    public Optional<Object> bindFrom(MqttV5BindingContext context, ArgumentConversionContext<Object> conversionContext) {
        return conversionService.convert(context.getProperties().getCorrelationData(), conversionContext); // <5>
    }
}
// end::clazz[]

File: test-suite-python/src/test/python/micronaut/mqtt/docs/custom/annotation/CorrelationAnnotationBinder.py
import java

# tag::imports[]
from jakarta.inject import Singleton
from micronaut.core.convert import ArgumentConversionContext, ConversionService
from micronaut.core.type import Argument
from micronaut.mqtt.bind import AnnotatedMqttBinder
from micronaut.mqtt.v5.bind import MqttV5BindingContext

from java.util import Optional
# end::imports[]

from micronaut.context.annotation import Requires

# TODO(python): java.type needed because the annotation type is returned to Java as a runtime java.lang.Class
# (AnnotatedMqttBinder.getAnnotationType()); the Python annotation function itself is not accepted
# ("Cannot convert '<function Correlation>' to Java type 'java.lang.Class'")
CorrelationClass = java.type("micronaut.mqtt.docs.custom.annotation.Correlation")


@Requires(property="spec.name", value="CorrelationSpec")
# tag::clazz[]
@Singleton  # <1>
class CorrelationAnnotationBinder(AnnotatedMqttBinder):  # <2>

    def __init__(self, conversion_service: ConversionService):  # <3>
        self.conversion_service = conversion_service

    def getAnnotationType(self):
        return CorrelationClass

    def bindTo(self, context: MqttV5BindingContext, value: object, argument: Argument) -> None:
        context.getProperties().setCorrelationData(value)  # <4>

    def bindFrom(self, context: MqttV5BindingContext, conversion_context: ArgumentConversionContext) -> Optional:
        return self.conversion_service.convert(context.getProperties().getCorrelationData(), conversion_context)  # <5>
# end::clazz[]
